gammapy/detect/test_statistics.py

```diff
@@ -72,30 +93,130 @@ def f_cash(x, counts, background, model):
+                              residual=False, morphology='Gaussian2D', width=0.3,
+                              *args, **kwargs):
+    """
+    Compute multiscale TS maps using compute_ts_map.
+
+    High level TS map computation using a multi gauss PSF kernel and assuming
+    a given source morphology. To optimize the performance the input data
+    can be sampled down when computing TS maps on larger scales.
+
+    Parameters
+    ----------
+    maps : `astropy.io.fits.HDUList`
+        HDU list containing the data. The list must contain the following HDU
+        extensions:
+            * 'On' -- Counts image
+            * 'Background' -- Background image
+            * 'Diffuse' -- Diffuse model image
+            * 'ExpGammaMap' -- Exposure image
+    psf_parameters : dict
+        Dict defining the multi gauss PSF parameters.
+        See `~gammapy.irf.multi_gauss_psf` for details.
+    scales : list ([0])
+        List of scales to use for TS map computation.
+    downsample : int (auto)
+        Down sampling factor. Can be set to 'auto' if the down sampling
+        factor should be chosen automatically.
+    residual : bool (False)
+        Compute a TS residual map.
+    morphology : str ('Gaussian2D')
+        Source morphology assumption. Either 'Gaussian2D' or 'Shell2D'.
+
+    Returns
+    -------
+    multiscale_result : list
+        List of `TSMapResult` objects.
+    """
+    BINSZ = abs(maps[0].header['CDELT1'])
+    shape = maps[0].data.shape
+    multiscale_result = []
+
+    for scale, factor in zip(scales, downsample):
+        logging.info('Computing (residual) TS map for scale {0} deg'.format(scale))
+
+        # Sample down and require that scale parameters is at least 5 pix
+        if downsample == 'auto':
+            downsample = np.select([scale < 5 * BINSZ, scale < 10 * BINSZ,
+                                    scale < 20 * BINSZ], [1, 2, 4], 8)
+        factor = int(downsample)
+        if factor == 1:
+            logging.info('No down sampling used.')
+            downsampled = False
+        else:
+            logging.info('Using down sampling factor of {0}'.format(factor))
+            downsampled = True
+
+        funcs = [np.nansum, np.mean, np.nansum, np.nansum, np.nansum]
+        maps_ = {}
+        for map_, func in zip(maps, funcs):
+            if downsampled:
+                maps_[map_.name] = downsample_2N(map_.data, factor, func,
+                                                 shape=shape_2N(shape))
+            else:
+                maps_[map_.name] = map_.data
+
+        # Set up PSF and source kernel
+        kernel = multi_gauss_psf_kernel(psf_parameters, BINSZ=BINSZ,
+                                        NEW_BINSZ=BINSZ * factor,
+                                        mode='oversample', x_size=21)
+
+        if scale > 0:
+            from astropy.convolution import convolve
+            sigma = scale / (BINSZ * factor)
+            if morphology == 'Gaussian2D':
```

Comment:
Always end such if -elif ... with an else and throw a `ValueError`.
If you don't the user will mistype the string and then your function will fail in weird hard-to-understand ways later.